File: projects/slick/lib/slick/view/file_interpreters.rb/md.rb

require "yaml"

Slick::View::FileInterpreter.register "md" do

    def interpret_file
        front_matter = {}
        body = File.read(file_path)
        
        if matches = body.match(/\A---(.*?)---(.*)\z/m)
            front_matter = YAML.load(matches[1])
            front_matter = {} if !front_matter.kind_of?(Hash)
            body = matches[2]
        end
        front_matter = front_matter.paramify
        body = grab { render_markdown(body) }
        
        front_matter[:title] ||= begin
            h1 = body.descendants.find{|node| node.type == 'h1'}
            h1 ? h1.text : nil
        end

        Slick::View.register(relative_file_path.sub(/\.md/i, '')).class_eval "
            def render
                render_view('_layout', #{front_matter.inspect}) do
                    echo #{body.to_s.dump}
                end
            end
        "
    end

end

File: packages/pinstripe/lib/view_file_importers/md.js

import { promisify } from 'util';
import { readFile } from 'fs';
import Yaml from 'js-yaml';

import { View } from '../view.js';

View.FileImporter.register('md', {
    importFile(){
        let extractParamsPromise;

        const { filePath, relativeFilePathWithoutExtension } = this;

        View.register(relativeFilePathWithoutExtension, {
            meta(){
                this.filePaths.push(filePath);
            },

            async render(){
                if(!extractParamsPromise){
                    extractParamsPromise = await this.extractParams();
                }
                const params = await extractParamsPromise;
                return this.renderView('_layout', params);
            },

            async extractParams(){
                const data = await promisify(readFile)(filePath, 'utf8');
                const [ frontMatter, body ] = this.extractFrontMatterAndBody(data);
                return { 
                    ...frontMatter, 
                    body: this.renderView('_content', {
                        body: this.renderMarkdown(body)
                    })
                };
            },

            extractFrontMatterAndBody(data){
                let frontMatter = {};
                let body = '';
                const matches = data.match(/^---+[\r\n]([\S\s]*?)[\r\n]---+([\S\s]*)$/)
                if(matches){
                    frontMatter = Yaml.load(matches[1]);
                    body = matches[2];
                } else {
                    body = data;
                }
                return [ frontMatter, body ];
            }
        });
    }
});
